Getting started with TypeScript example (#1681)
commercetools/nodejs: docs/sdk/api/typescriptSdk.md

</script>
```

## Getting started
This tutorial will show you how to use the middlewares in this **[commercetools TypeScript SDK](https://github.com/commercetools/commercetools-sdk-typescript/)** to get a simple commercetools JavaScript app running.

### Create a API client
[Create API client](https://docs.commercetools.com/tutorials/getting-started#creating-an-api-client) from Merchant Center. If you do not have account [follow the steps to create a free trial account](https://docs.commercetools.com/tutorials/getting-started#first-steps). 
In this guide we’ll be calling a method of commercetools API using TypeScript SDK to get the settings of a commercetools project. The commercetools API is the foundation of the commercetools Platform, and almost every commercetools client app uses it. Aside from getting project information, the commercetools API allows clients to call methods that can be used for everything from creating products to updating an order’s status. Before we can call any methods, we need to configure our new app to obtain an access token.

### Set up your local project

> If you don’t already have a project, let’s create a new one. In an empty directory, you can initialize a new project using the following command:

Reviewer: Can we say "Node.js project" here? I just want to be clear it cannot be confused with the commercetools project that should be existing at this step for sure.

Author: Yes

Run the program using the following command:
```
$ node project.js
Getting started with commercetools Typescript SDK
```
If you see the same output as above, we’re ready to start.

### Get commercetools project settings with the commercetools API
In this guide we’ll get project settings information. Re-open `project.js` and add the following code:
```js
(async () => {
    try {
        await apiRoot.withProjectKey({projectKey}).get().execute()
            .then(data => {
                console.log('Project information --->', data);
            })
            .catch(error => {
                console.log('ERROR --->', error);
            })
    } catch (error) {
        console.log('ERROR --->', error);
    }
})();
```
Then client will **execute** get project information request from `apiRoot` using **TypeScript SDK**.
Run the program. The output should look like the following if the request is successful:
```
$ node project.js
Getting started with commercetools Typescript SDK
Project information ---> {
  body: {

>     <projectData>

Reviewer: ```suggestion
    key: '<project_key>',
    name: '<project_name>',
    countries: [ ... ],
    currencies: [ ... ],
    languages: [ ... ],
    createdAt: '...',
    ...
```

const { createHttpMiddleware } = require('@commercetools/sdk-middleware-http')
const { createApiBuilderFromCtpClient } = require("@commercetools/typescript-sdk");
  
const fetch = require('node-fetch')
require('dotenv').config()

console.log('Getting started with commercetools Typescript SDK');
```

Back at the command line, run the program using the following command:
```
$ node project.js
Getting started with commercetools Typescript SDK
```
If you see the same output as above, we’re ready to start.

### Get commercetools project settings with the commercetools API
In this guide we’ll get project settings information. We’ll also follow the best practice of keeping secrets outside of your code (do not hardcode sensitive data).

Store the **client id** and **secret** in a new environment variable. Create a new file called `.env` in this directory and add the following code: 

```
ADMIN_CLIENT_ID=<your_admin_client_id>
ADMIN_CLIENT_SECRET=<your_admin_secret_id>
```
Replace the values with your client id and client secret that you copied earlier.

Re-open `project.js` and add the following code:
```js
const { 
    ADMIN_CLIENT_ID,
    ADMIN_CLIENT_SECRET,
} = process.env;

> const projectKey = '<your_project_key>'

Reviewer: Could we read this from the `.env` file as well? Then this part does not have to be edited by the reader.

// Create a httpMiddleware for the your project AUTH URL
const authMiddleware = createAuthMiddlewareForClientCredentialsFlow({

>     host: '<your_auth_url>',

Reviewer: Could we read this from the `.env` file as well? Then this part does not have to be edited by the reader.

    projectKey,
    credentials: {
        clientId: ADMIN_CLIENT_ID,
        clientSecret: ADMIN_CLIENT_SECRET,
    },

>     scopes: ['<your_client_scope>'],

Reviewer: Could we read this from the `.env` file as well? Then this part does not have to be edited by the reader.

    fetch,
})

// Create a httpMiddleware for the your project API URL
const httpMiddleware = createHttpMiddleware({

>     host: '<your_api_url>',

Reviewer: Could we read this from the `.env` file as well? Then this part does not have to be edited by the reader.

</script>
```

## Getting started
This tutorial will show you how to use the middlewares in this **[commercetools TypeScript SDK](https://github.com/commercetools/commercetools-sdk-typescript/)** to get a simple commercetools JavaScript app running.

### Create a API client

> [Create API client](https://docs.commercetools.com/tutorials/getting-started#creating-an-api-client) from Merchant Center. If you do not have account [follow the steps to create a free trial account](https://docs.commercetools.com/tutorials/getting-started#first-steps). 

Reviewer 2: ```suggestion
[Create an API client](https://docs.commercetools.com/tutorials/getting-started#creating-an-api-client) from the Merchant Center.

If you do not have account [follow these steps to create a free trial account](https://docs.commercetools.com/tutorials/getting-started#first-steps). 

When you have set up your API Client, ensure that you save your **project_key**, **client_id**, **secret**, **scope**, **API URL**, and **Auth URL**. We recommend using the drop-down list to save your API Client details as an **Environment Variables (.env)** file.
```

> In this guide we’ll be calling a method of commercetools API using TypeScript SDK to get the settings of a commercetools project. The commercetools API is the foundation of the commercetools Platform, and almost every commercetools client app uses it. Aside from getting project information, the commercetools API allows clients to call methods that can be used for everything from creating products to updating an order’s status. Before we can call any methods, we need to configure our new app to obtain an access token.

Reviewer 2: ```suggestion

In this guide we call a method of the commercetools API using the TypeScript SDK to get the settings of a commercetools project.

The commercetools API is the foundation of the commercetools Platform, and almost every commercetools client app uses it.

Aside from retrieving project information, the commercetools API allows clients to call methods that can be used for everything from creating products to updating an order’s status.

Before we can call any methods, we need to configure a new app to obtain an access token.
```

    host: CTP_AUTH_URL,
    projectKey,
    credentials: {
      clientId: CTP_CLIENT_ID,
      clientSecret: CTP_CLIENT_SECRET,
    },
    scopes: [CTP_SCOPES],
    fetch,
  })
  const httpMiddleware = createHttpMiddleware({
    host: CTP_API_URL,
    fetch,
  })
  const client = createClient({
    middlewares: [authMiddleware, httpMiddleware],
  })

  // Create an API root from API builder of commercetools platform client
  const apiRoot = createApiBuilderFromCtpClient(client);

console.log('Getting started with commercetools Typescript SDK');
```

You must now install the following packages: `@commercetools/sdk-client`, `@commercetools/sdk-middleware-auth`, `@commercetools/sdk-middleware-http`, `@commercetools/typescript-sdk` and `dotenv`.

You can do so by using the following command:

```
npm install @commercetools/sdk-client @commercetools/sdk-middleware-auth @commercetools/sdk-middleware-http @commercetools/typescript-sdk dotenv
```

This code creates a **client**, which uses `authMiddleware` and `httpMiddleware`. The `httpMiddleware` reads the `clientId` and `clientSecret` from the `.env` file.

Run the program using the following command:
```node project.js```

> You should see the following message in your output:

Reviewer 2: ```suggestion

You should see the following message in your output:
```

You should see the following message in your output:
```Getting started with commercetools Typescript SDK```

> If you have the same output as above, then you have set up everything correctly.

Reviewer 2: ```suggestion

If you have the same output as above, then you have set up everything correctly.
```

> If you have the same output as above, then you have set up everything correctly.

Reviewer 2: ```suggestion

If you have the same output as above, then you have set up everything correctly.
```

When you have set up your API Client, ensure that you save your **project_key**, **client_id**, **secret**, **scope**, **API URL**, and **Auth URL**. We recommend using the drop-down list to save your API Client details as an **Environment Variables (.env)** file.

In this guide we call a method of the commercetools API using the TypeScript SDK to get the settings of a commercetools project.

The commercetools API is the foundation of the commercetools Platform, and almost every commercetools client app uses it.

Aside from retrieving project information, the commercetools API allows clients to call methods that can be used for everything from creating products to updating an order’s status.

Before we can call any methods, we need to configure a new app to obtain an access token.

### Set up your local project
1. Open a terminal window.
2. Create a new directory for the project.
3. Open the new directory and create a new project using the following command: ```npm init -y```
4. The project will be created and a package.json file will be created in the directory.

> 5. Enter the following command to open this folder in your code editor ```code .````

Reviewer 2: ```suggestion
5. Enter the following command to open this folder in your code editor: ```code .```
```

> ```

Reviewer 2: ```suggestion

```

    },
    scopes: [CTP_SCOPES],
    fetch,
  })
  const httpMiddleware = createHttpMiddleware({
    host: CTP_API_URL,
    fetch,
  })
  const client = createClient({
    middlewares: [authMiddleware, httpMiddleware],
  })

  // Create an API root from API builder of commercetools platform client
  const apiRoot = createApiBuilderFromCtpClient(client);

console.log('Getting started with commercetools Typescript SDK');
```

You must now install the following packages: `@commercetools/sdk-client`, `@commercetools/sdk-middleware-auth`, `@commercetools/sdk-middleware-http`, `@commercetools/typescript-sdk` and `dotenv`.

You can do so by using the following command:

```
npm install @commercetools/sdk-client @commercetools/sdk-middleware-auth @commercetools/sdk-middleware-http @commercetools/typescript-sdk dotenv
```

This code creates a **client**, which uses `authMiddleware` and `httpMiddleware`. The `httpMiddleware` reads the `clientId` and `clientSecret` from the `.env` file.

Run the program using the following command:
```node project.js```

You should see the following message in your output:
```Getting started with commercetools Typescript SDK

> ```

Reviewer 2: ```suggestion

```

      clientId: CTP_CLIENT_ID,
      clientSecret: CTP_CLIENT_SECRET,
    },
    scopes: [CTP_SCOPES],
    fetch,
  })
  const httpMiddleware = createHttpMiddleware({
    host: CTP_API_URL,
    fetch,
  })
  const client = createClient({
    middlewares: [authMiddleware, httpMiddleware],
  })

  // Create an API root from API builder of commercetools platform client
  const apiRoot = createApiBuilderFromCtpClient(client);

console.log('Getting started with commercetools Typescript SDK');
```

You must now install the following packages: `@commercetools/sdk-client`, `@commercetools/sdk-middleware-auth`, `@commercetools/sdk-middleware-http`, `@commercetools/typescript-sdk` and `dotenv`.

You can do so by using the following command:

```
npm install @commercetools/sdk-client @commercetools/sdk-middleware-auth @commercetools/sdk-middleware-http @commercetools/typescript-sdk dotenv
```

This code creates a **client**, which uses `authMiddleware` and `httpMiddleware`. The `httpMiddleware` reads the `clientId` and `clientSecret` from the `.env` file.

Run the program using the following command:
```node project.js```

You should see the following message in your output:

> ```Getting started with commercetools Typescript SDK```

Reviewer 2: ```suggestion

```Getting started with commercetools Typescript SDK```
```
